### PR TITLE
DATAUP-280 Migrate tests to github actions

### PR DESCRIPTION
**Pull request by eamahanna**
# Description of PR purpose/changes

This PR implements a few major changes across a few tickets
* DATAUP-280 explore narrative test migration to github actions  
I know this ticket was meant to just do an estimate of effort for that migration, but the work was nearly completed a few months back, and this mostly finalizes it. Other adjustments can be made in proceeding PRs.
* DATAUP-232 Implement the local integration testing
* DATAUP-233 Implement integration testing in Travis-CI
These two PRs revolve around implementing integration testing using webdriver.io in this repo. This work was done to work both locally with `make test-integration` and on Github actions. Note that an environment variable `KBASE_TEST_TOKEN` is required. This needs to be set up locally (just use a developer token), but is done via Github secrets in the Narrative repo for the testing action. Also note that for integration tests to work via Github actions, only a branch of this repo - not a fork - can be used.

Finally, due to some technical problems in the action, reporting to coveralls.io has been replaced with reporting to codecov.io. I'm up for conversation about this, as I don't want to make that selection by fiat. I just had trouble getting the combined results of both Python and Javascript tests to get submitted to coveralls via Github action, but the supplied codecov process was much cleaner. Codecov is also used in other KBase core service repos like [Workspace](https://github.com/kbase/workspace_deluxe) and [Execution Engine](https://github.com/kbase/execution_engine2).

Currently, tests run when making a PR against the `develop`, `master`, or `truss` branches.

Regarding integration tests, these run through the `run_tests.py` script that the unit tests use. When running with `make test-integration`, that script starts up a narrative interface, then passes the base url to an environment variable that the test runner makes use of. You can also do this manually from a shell with:
```
> kbase-narrative
> npx wdio test/integration/wdio.conf.js
```

# Jira Ticket / Issue #
e.g. https://kbase-jira.atlassian.net/browse/DATAUP-X
- [ ] Added the Jira Ticket to the title of the PR e.g. (DATAUP-69 Adds a PR template)

# Testing Instructions
* Details for how to test the PR: 

Pull the repo, do a fresh npm install, add your token to the environment with `export KBASE_TEST_TOKEN=my_dev_token` and run `make test` for the whole test suite. Alternately, just run `make test-integration`

- [ ] Tests pass in Github actions and locally 
- [ ] Changes available by spinning up a local narrative and navigating to _X_ to see _Y_

# Dev Checklist:

- [ ] My code follows the guidelines at https://sites.google.com/truss.works/kbasetruss/development
- [ ] I have performed a self-review of my own code
- [ ] I have commented my code, particularly in hard-to-understand areas
- [ ] I have made corresponding changes to the documentation
- [ ] My changes generate no new warnings
- [ ] I have added tests that prove my fix is effective or that my feature works
- [ ] New and existing unit tests pass locally with my changes
- [ ] Any dependent changes have been merged and published in downstream modules
- [ ] I have run Black and Flake8 on changed Python Code manually or with git precommit (and the travis build passes)

### CODE REVIEW

**Comment by briehl** [comment]
Working my way through some docs changes, and a little bit around token management, then should be ready by Monday morning. If Travis-CI wants to cooperate.

**Comment by briehl** [comment]
Codacy is mainly complaining about the test runner calls executing "untrusted" input. I'm not sure if that really applies here or not, and their suggestions aren't helpful.

File: .github/workflows/test.yml [hunk 1]
```diff
@@ -0,0 +1,63 @@
+name: CI-testing
+
+on:
+  push:
+    branches: [ develop, master, truss ]
+  pull_request:
+    branches: [ develop, master, truss ]
```
Comment (ialarmedalien) [inline .github/workflows/test.yml:7]: unless there is a particular reason for restricting the test runs to these branches, I would set the action to run on push and pull request on any branch, i.e.

```yaml
on:
  [push, pull_request]
jobs:
  ...
```

Comment (briehl) [reply .github/workflows/test.yml]: Done. Thanks for the suggestion.

File: .travis.yml [hunk 2]
```diff
@@ -13,6 +13,7 @@ env:
     - TRAVIS_NODE_VERSION="10.10.0"
     - secure: "JPdkdywgQbUUixuDuATMXZHWpWImziRoGKd1tCjYd3+0lamLBlldmjhsZ+Bp7ZabKg2ExQZareGDp34cJdOwWBGb2gg0/emjy4UkWkbDH28zyoWNK1SFp6OTY8AeY/icPXKw3MBgKoXG+hemfcxhYc8rUNgUt7V2fA0JSAkCi4w="
     - secure: "BrGlbfGLxZYCynh7LeIFyQeiTx4YX6fdMYK2UceVLoAFjOml9mvtJY5i4Oafc2iBDPQ0wAP/eIzZOTcddwc+12o1S0EAq3zZb5z+MOwRKjsiC+BGUXNzAGLaeJqFTC5Sn+vvqsaDGaR3jrdQ8APUL7XapuLUAkof7vqc1SPsn3I="
+    - secure: "YH4s6huZYW34XFlKDAg7yC8VGUjIiird3gDKOmHyhlPVoQa75tll3iqVBBxNFcj+5CfzlCU+ZZK/JTvRoLJ7+QBPrRm7eZ6+s44JaJK+eAz12AJKdEfZyp+blMQ26Y8uOz5AZQhnRsoYfbqDK3p/T8FW1MAKdvE+8nHDCdAh26k="
```
Comment (eamahanna) [inline .travis.yml:16]: Does this travis key meed to be here given we are going with github actions?

File: docs/testing.md [hunk 3]
```diff
@@ -67,7 +68,18 @@ This just needs the path to the token file (with pre-pended slash), such as `"/t
 
 ### Testing with Travis-CI and Coveralls
 
-These tests are run (without credentials) automatically on a pull request to the Narrative Github repo. These are currently run through [Travis-CI](https://travis-ci.org/) and the coverage reported with [Coveralls](https://coveralls.io/). There should be nothing you need to do to make this work.
+These tests are run automatically on a pull request to the Narrative Github repo. These are currently run through [Travis-CI](https://travis-ci.org/) and the coverage reported with [Coveralls](https://coveralls.io/).
```
Comment (eamahanna) [inline docs/testing.md:71]: Should this be github actions?

File: .travis.yml (hunk 2 above)
Comment (briehl) [reply .travis.yml:16]: Nope. In fact that whole file can go away. Should I just do that now? I was going to do that in a separate PR, but if it's appropriate here, I'll nuke it.

File: docs/testing.md (hunk 3 above)
Comment (briehl) [reply docs/testing.md]: Yes, and Codecov. Updated that paragraph and some other places in that document.

File: .travis.yml (hunk 2 above)
Comment (eamahanna) [reply .travis.yml:16]: I am fine putting it in a separate ticket or in this PR. I think we have DATAUP-234 to remove old code, but its up to you.

File: Makefile [hunk 4]
```diff
@@ -49,20 +49,20 @@ test-backend:
...
 
-# test-frontend-e2e should use Selenium to perform an end-
-# to-end test of the front end components, with a running
-# Narrative system.
-test-frontend-e2e:
-	@echo "running frontend end-to-end tests"
-	cd $(FRONTEND_TEST_DIR)
+test-integration:
+	@echo "running integration tests"
+	python test/unit/run_tests.py -i
```
Comment (eamahanna) [inline Makefile:65]: To execute the integration tests locally, will we need to supply a token?

File: docs/testing.md [hunk 5]
```diff
@@ -65,9 +66,13 @@ This just needs the path to the token file (with pre-pended slash), such as `"/t
...
 
-### Testing with Travis-CI and Coveralls
+### Testing with Github Actions and Codecov
 
-These tests are run (without credentials) automatically on a pull request to the Narrative Github repo. These are currently run through [Travis-CI](https://travis-ci.org/) and the coverage reported with [Coveralls](https://coveralls.io/). There should be nothing you need to do to make this work.
+These tests are run automatically on a pull request to the Narrative Github repo. These are currently run through [Github Actions](https://docs.github.com/en/free-pro-team@latest/actions) and the coverage reported with [Codecov](https://codecov.io/).
+
+Unit tests are automatically run without credentials, skipping various tests that are, really, more like integration tests.
+
+The integration tests that run with webdriver.io do require an authentication token. This is the `NARRATIVE_TEST_TOKEN` Github secret in the Narrative repo. It will become available in the test environment as `KBASE_TEST_TOKEN`, which is the variable that the `wdio.conf.js` file looks for.
```
Comment (eamahanna) [inline docs/testing.md:75]: Can you add a quick description on how to run the tests locally?

File: docs/testing.md [hunk 6]
```diff
@@ -42,6 +42,7 @@ Then, simply run (from the narrative root directory) `make test`.
 This calls a few subcommands, and those can be run independently for specific uses:
 
 - `make test-frontend-unit` will run only the unit tests on the frontend (i.e. those with the Karma runner)
+- `make test-integration` will run the frontend integration tests that make use of webdriver.io to simulate the browser on a locally instantiated Narrative, but running against live KBase services.
```
Comment (ialarmedalien) [inline docs/testing.md:45]: Do you want to add in `make test-frontend`, which runs the integration and frontend unit tests?

**Review by ialarmedalien — APPROVED** [review]
A few comments but no blockers.

File: docs/testing.md (hunk 6 above)
Comment (briehl) [reply docs/testing.md:45]: Yup, I'll add that to the docs.

File: .travis.yml (hunk 2 above)
Comment (briehl) [reply .travis.yml:16]: I'll just use that other ticket to remove the travis stuff and update that PR. Thanks for the feedback!

File: docs/testing.md (hunk 6 above)
Comment (briehl) [reply docs/testing.md]: Added.

File: docs/testing.md (hunk 5 above)
Comment (briehl) [reply docs/testing.md:84]: I updated the docs with local test run info. I think that's in the right place and helpful? It's near the top with the other make directives.

File: Makefile (hunk 4 above)
Comment (briehl) [reply Makefile:65]: Yes, I updated the docs to say so, and where to put it.

**Comment by briehl** [comment]
I think this one is ready to go if anyone wants to give the docs one last check-over.

**Review by ialarmedalien — APPROVED** [review]
This looks good.

**Pull request merged**